File: apps/web/src/api/timetable.ts
import client from './client';

export interface RoomData {
  id: string;
  tenant_id: string;
  name: string;
  campus_id: string | null;
  capacity: number | null;
  equipment: string[];
  created_at: string;
}

export interface SessionData {
  id: string;
  tenant_id: string;
  academic_year_id: string | null;
  group_id: string;
  subject_id: string;
  teacher_id: string;
  room_id: string | null;
  day_of_week: number;
  start_time: string;
  end_time: string;
  recurrence: string;
  effective_from: string | null;
  effective_until: string | null;
  status: string;
  created_at: string;
}

export interface SessionExceptionData {
  id: string;
  session_id: string;
  date: string;
  exception_type: string;
  substitute_teacher_id: string | null;
  new_room_id: string | null;
  new_start_time: string | null;
  new_end_time: string | null;
  reason: string | null;
  created_by: string;
  created_at: string;
}

export function getSessions(params: { group_id?: string; teacher_id?: string; day_of_week?: number } = {}) {
  return client.get<SessionData[]>('/v1/timetable/sessions', { params });
}

export function createSession(data: Omit<SessionData, 'id' | 'tenant_id' | 'status' | 'created_at'>) {
  return client.post<SessionData>('/v1/timetable/sessions', data);
}

export function updateSession(id: string, data: Partial<SessionData>) {
  return client.patch<SessionData>(`/v1/timetable/sessions/${id}`, data);
}

export function deleteSession(id: string) {
  return client.delete(`/v1/timetable/sessions/${id}`);
}

export function getRooms() {
  return client.get<RoomData[]>('/v1/timetable/rooms');
}

export function createRoom(data: { name: string; campus_id?: string; capacity?: number; equipment?: string[] }) {
  return client.post<RoomData>('/v1/timetable/rooms', data);
}

export function createSessionException(data: {
  session_id: string;
  date: string;
  exception_type: string;
  substitute_teacher_id?: string;
  new_room_id?: string;
  new_start_time?: string;
  new_end_time?: string;
  reason?: string;
}) {
  return client.post<SessionExceptionData>('/v1/timetable/session-exceptions', data);
}

File: apps/api/app/db/models/timetable.py
import uuid
from datetime import datetime

from sqlalchemy import Column, Date, DateTime, Enum, ForeignKey, Integer, String, Text, Time
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import relationship

from app.db.base import Base, TenantMixin


class Room(Base, TenantMixin):
    __tablename__ = "rooms"

    campus_id = Column(UUID(as_uuid=True), ForeignKey("campuses.id"), nullable=True)
    name = Column(String(255), nullable=False)
    capacity = Column(Integer, nullable=True)
    equipment = Column(ARRAY(Text), default=list)


class Session(Base, TenantMixin):
    __tablename__ = "sessions"

    academic_year_id = Column(UUID(as_uuid=True), ForeignKey("academic_years.id"), nullable=True)
    group_id = Column(UUID(as_uuid=True), ForeignKey("groups.id"), nullable=False)
    subject_id = Column(UUID(as_uuid=True), ForeignKey("subjects.id"), nullable=False)
    teacher_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    room_id = Column(UUID(as_uuid=True), ForeignKey("rooms.id"), nullable=True)
    day_of_week = Column(Integer, nullable=False)  # 0=Monday ... 6=Sunday
    start_time = Column(Time, nullable=False)
    end_time = Column(Time, nullable=False)
    recurrence = Column(String(20), nullable=False, default="weekly")  # weekly|biweekly|custom
    effective_from = Column(Date, nullable=True)
    effective_until = Column(Date, nullable=True)
    status = Column(String(20), nullable=False, default="active")  # active|cancelled|substituted

    exceptions = relationship("SessionException", back_populates="session", cascade="all, delete-orphan")


class SessionException(Base):
    __tablename__ = "session_exceptions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(UUID(as_uuid=True), ForeignKey("sessions.id"), nullable=False)
    date = Column(Date, nullable=False)
    exception_type = Column(String(20), nullable=False)  # cancelled|substituted|room_change|time_change
    substitute_teacher_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    new_room_id = Column(UUID(as_uuid=True), ForeignKey("rooms.id"), nullable=True)
    new_start_time = Column(Time, nullable=True)
    new_end_time = Column(Time, nullable=True)
    reason = Column(Text, nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    session = relationship("Session", back_populates="exceptions")
